map/mwm_url.cpp

```diff
@@ -19,155 +19,234 @@
+  }
+
+  ParsingResult const res = Parse(url_scheme::Uri(url));
+  m_isValid = res != ParsingResult::Incorrect;
+  return res;
 }
 
-bool ParsedMapApi::Parse(Uri const & uri)
+ParsedMapApi::ParsingResult ParsedMapApi::Parse(Uri const & uri)
 {
   string const & scheme = uri.GetScheme();
-  if ((scheme != "mapswithme" && scheme != "mwm") || uri.GetPath() != "map")
-    return false;
+  string const & path = uri.GetPath();
+  bool const isRoutePath = path == "route";
+  if ((scheme != "mapswithme" && scheme != "mwm" && scheme != "mapsme") ||
+      (path != "map" && !isRoutePath))
+  {
+    return ParsingResult::Incorrect;
+  }
 
-  ASSERT(m_bmManager != nullptr, ());
-  UserMarkControllerGuard guard(*m_bmManager, UserMarkType::API_MARK);
+  if (isRoutePath)
+  {
+    m_routePoints.clear();
+    vector<string> pattern{kSourceLatLon, kSourceName, kDestLatLon, kDestName, kRouteType};
+    if (!uri.ForEachKeyValue(bind(&ParsedMapApi::RouteKeyValue, this, _1, _2, ref(pattern))))
+      return ParsingResult::Incorrect;
+
+    if (pattern.size() != 0)
+      return ParsingResult::Incorrect;
+
+    if (m_routePoints.size() != 2)
+    {
+      ASSERT(false, ());
+      return ParsingResult::Incorrect;
+    }
 
+    return ParsingResult::Route;
+  }
+
+  ASSERT(m_bmManager != nullptr, ());
   vector<ApiPoint> points;
-  uri.ForEachKeyValue(bind(&ParsedMapApi::AddKeyValue, this, _1, _2, ref(points)));
-  points.erase(remove_if(points.begin(), points.end(), &IsInvalidApiPoint), points.end());
+  if (!uri.ForEachKeyValue(bind(&ParsedMapApi::AddKeyValue, this, _1, _2, ref(points))))
+    return ParsingResult::Incorrect;
+
+  if (points.empty())
+    return ParsingResult::Incorrect;
 
-  for (size_t i = 0; i < points.size(); ++i)
+  UserMarkControllerGuard guard(*m_bmManager, UserMarkType::API_MARK);
+  for (auto const & p : points)
   {
-    ApiPoint const & p = points[i];
     m2::PointD glPoint(MercatorBounds::FromLatLon(p.m_lat, p.m_lon));
     ApiMarkPoint * mark = static_cast<ApiMarkPoint *>(guard.m_controller.CreateUserMark(glPoint));
     mark->SetName(p.m_name);
     mark->SetID(p.m_id);
     mark->SetStyle(style::GetSupportedStyle(p.m_style, p.m_name, ""));
   }
 
-  return true;
+  return ParsingResult::Map;
 }
 
-void ParsedMapApi::AddKeyValue(string key, string const & value, vector<ApiPoint> & points)
+bool ParsedMapApi::RouteKeyValue(string key, string const & value, vector<string> & pattern)
 {
-  strings::AsciiToLower(key);
+  if (key != pattern.front())
+    return false;
 
-  if (key == "ll")
+  if (key == kSourceLatLon || key == kDestLatLon)
   {
-    points.push_back(ApiPoint());
-    points.back().m_lat = INVALID_LAT_VALUE;
+    double lat = 0.0;
+    double lon = 0.0;
+    if (!ParseLatLon(key, value, lat, lon))
+      return false;
 
-    size_t const firstComma = value.find(',');
-    if (firstComma == string::npos)
+    RoutePoint p;
+    p.m_org = MercatorBounds::FromLatLon(lat, lon);
+    m_routePoints.push_back(p);
```

Comment:
For example, imagine API request, where fields are in the following order:
```
dest lat lon, source name, source lat lon, dest name
```
What'll happen in the case?